// multi-avatar/lib/provider/vk.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _anyPromise = require('any-promise');

var _anyPromise2 = _interopRequireDefault(_anyPromise);

var _nodeFetch = require('node-fetch');

var _nodeFetch2 = _interopRequireDefault(_nodeFetch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

_nodeFetch2.default.Promise = _anyPromise2.default;

module.exports = function () {
  function _class(vkId) {
    var timeout = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 5000;

    _classCallCheck(this, _class);

    this.vkId = vkId;
    this.timeout = timeout;

    this.pictureUrls = null;
    this.error = false;
  }

  _createClass(_class, [{
    key: 'providerName',
    value: function providerName() {
      return 'vk';
    }
  }, {
    key: 'getImageSize',
    value: function getImageSize(size) {
      if (size <= 50) return 0;

      if (size <= 100) return 1;

      return 2;
    }
  }, {
    key: 'avatar',
    value: function avatar(size) {
      var _this = this;

      return new _anyPromise2.default(function (resolve, reject) {
        if (!_this.vkId) return reject(new Error('Vk user ID is not specified'));
        if (!!_this.error) return reject(_this.error);

        if (!_this.pictureUrls) {
          (0, _nodeFetch2.default)('https://api.vk.com/method/users.get?user_id=' + _this.vkId + '&v=5.8&fields=photo_50,photo_100,photo_200').then(function (res) {
            return res.json();
          }).then(function (data) {
            if (data.error) {
              if (data.error.error_code == 100) {
                _this.error = new Error(data.error.error_msg);
              }
              return reject(new Error(data.error.error_msg));
            }

            _this.pictureUrls = [data.response[0].photo_50, data.response[0].photo_100, data.response[0].photo_200];
            return resolve();
          });
        } else {
          return resolve();
        }
      }).then(function () {
        return _this.pictureUrls[_this.getImageSize(size)];
      });
    }
  }]);

  return _class;
}();
